### ISSUE TITLE
Consider adding optional template argument for custom "map" implementation and keep the default

### ISSUE THREAD

**Issue by Alexander-Shukaev** [jimporter/memo#1]
As per subject.

**Comment by jimporter** [comment]
Indeed. When most of this code was written, `std::map` was the only (standard) associative container with heterogeneous lookup, but as of C++20, `std::unordered_map` allows this too. (And there are the `std::pmr` types in C++17, though I'm not sure they'd be especially useful here.) And, of course, it'd be nice to be able to use some kind of LRU cache.
